MAINT: Apply assorted ruff/flake8-simplify rules (SIM)

DimitriPapadopoulos opened the pull request:
Merging #27300 first will help with residual linter errors.

DimitriPapadopoulos committed:
MAINT: Apply ruff/flake8-simplify rule SIM101

SIM101 Multiple `isinstance` calls, merge into a single call

DimitriPapadopoulos committed:
MAINT: Apply ruff/flake8-simplify rule SIM401

SIM401 Use `.get(attrib, ())` instead of an `if` block

DimitriPapadopoulos committed:
MAINT: Apply ruff/flake8-simplify rule SIM910

SIM910 Use `.get('out')` instead of `.get('out', None)`

charris commented:
The circleci failures are unrelated and fixed in main. I'd like to see the lint failures fixed before going to much further.

DimitriPapadopoulos commented:
The lint failures are in distutils. I am working on getting `tools/linter.py` to skip `tools/linter.py` as instructed in `ruff.toml`. See #28015.

charris commented on numpy/_core/numerictypes.py:
```diff
@@ -223,10 +223,7 @@ def issctype(rep):
-            return True
-        else:
-            return False
+        return res and res != object_
```
This isn't the same, the return type changes, it was bool.

seberg replied:
Agreed, let's just skip rule 103 at least until it is fixed (although, once you add a bool on top, I lean towards most of this code is being better without the change).

114 (merging ifs with identical branches) isn't quite as nice to me as I would have expected, but if @charris likes it, happy to go with it.

(401 and 910, the dict changes are great IMO)

DimitriPapadopoulos replied:
This was a manual change. I keep forgetting that unlike `&&` in C, the result of `and` in Python is not necessarily a bool.

DimitriPapadopoulos replied:
Overall, I have mixed feelings about the `SIM` rules as well. Many of these rules may improve or worsen readability, depending on the original code. It's not a good idea to apply them automatically. Let me try to remove the more debatable rules.

charris commented on numpy/_core/tests/_natype.py:
```diff
@@ -14,10 +14,8 @@ def _create_binary_propagating_op(name, is_divmod=False):
-            or isinstance(other, np.ndarray)
-            and not other.shape
+            or isinstance(other, (str, bytes, numbers.Number, np.bool))
+            or isinstance(other, np.ndarray) and not other.shape
```
I think the previous was more readable.

DimitriPapadopoulos replied:
Not sure about it. Calling the `isinstance` function twice doesn't seem more readable to me.

seberg replied:
Would we apply these automatically (Maybe even as a `pre-commit`) in a future?
In that case, I happy to just do it.  Otherwise it is matter of opinion (the old code grouped things nicely).

If we enforce/use more auto-formatting, I suspect that one add comments to avoid it if desired (I don't think this instance matters much):
```
        or isinstance(other, (str, bytes))  # str-like
```
which is also true for some of the other simplification rules.

DimitriPapadopoulos replied:
I would be for applying some micro-optimisations automatically using the ruff linter. They would still be micro-optimisations of course, but I think that's OK if they're applied in a systematic and automatic way.

This would be such a micro-optimisation applied by the ruff linter: one single `isinstance` call instead of multiple ones - more than a formatting issue applied by the ruff formatter.

charris commented on numpy/f2py/crackfortran.py:
```diff
@@ -1933,9 +1929,7 @@ def setattrspec(decl, attr, force=0):
-    if attr == 'static' and 'automatic' not in decl['attrspec']:
-        decl['attrspec'].append(attr)
-    elif attr == 'automatic' and 'static' not in decl['attrspec']:
+    if attr == 'static' and 'automatic' not in decl['attrspec'] or attr == 'automatic' and 'static' not in decl['attrspec']:
```
I think the original is more readable.

DimitriPapadopoulos replied:
Agreed. Perhaps temporary variables with meaningful names would be even easier to read.

charris commented on numpy/f2py/crackfortran.py:
```diff
@@ -2271,9 +2265,8 @@ def analyzebody(block, args, tab=''):
-           not b['body'] and not b.get('implementedby'):
-            if 'f2pyenhancements' not in b:
-                continue
+           not b['body'] and not b.get('implementedby') and 'f2pyenhancements' not in b:
```
Neither is very readable here, I'd put more of the bits in separate lines.

DimitriPapadopoulos replied:
Agreed, but if numpy starts using a formatter some day, the formatter will unfold the line.

charris commented on numpy/f2py/crackfortran.py:
```diff
@@ -2892,17 +2882,16 @@ def compute_deps(v, deps):
+            elif isstring(vars[n]) and 'charselector' in vars[n]:
+                if '*' in vars[n]['charselector']:
+                    length = _eval_length(vars[n]['charselector']['*'],
+                                          params)
```
No need to break this line.

DimitriPapadopoulos replied:
I've just run ruff with this specific rule automatically. A formatter would unfold the line, but the linter doesn't.

charris commented on numpy/lib/_arraypad_impl.py:
```diff
@@ -848,7 +848,7 @@ def pad(array, pad_width, mode='constant', **kwargs):
     elif mode in stat_functions:
         func = stat_functions[mode]
-        length = kwargs.get("stat_length", None)
+        length = kwargs.get("stat_length")
```
It could be argued that the original was better because it doesn't require the reader to remember the default.

DimitriPapadopoulos replied:
The `.get()` function is quite basic. If the reader doesn't know that it returns `None` by default, then he probably doesn't know what the function is about and would have to look it into the documentation any way.

charris commented on numpy/lib/_npyio_impl.py:
```diff
@@ -1604,9 +1603,7 @@ def first_write(self, v):
-            elif iscomplex_X and n_fmt_chars != (2 * ncol):
-                raise error
-            elif ((not iscomplex_X) and n_fmt_chars != ncol):
+            elif iscomplex_X and n_fmt_chars != (2 * ncol) or ((not iscomplex_X) and n_fmt_chars != ncol):
```
The original is easier to read.

DimitriPapadopoulos replied:
Perhaps temporary variables with meaningful names would make it easier to read. I have to agree this looks less readable to me.

charris commented on numpy/lib/_shape_base_impl.py:
```diff
@@ -728,9 +728,7 @@ def dstack(tup):
-        if _nx.ndim(sub_arys[i]) == 0:
-            sub_arys[i] = _nx.empty(0, dtype=sub_arys[i].dtype)
-        elif _nx.sometrue(_nx.equal(_nx.shape(sub_arys[i]), 0)):
+        if _nx.ndim(sub_arys[i]) == 0 or _nx.sometrue(_nx.equal(_nx.shape(sub_arys[i]), 0)):
```
I don't mind the combination, but this would read better if broken across two lines.

DimitriPapadopoulos replied:
I could do that, but if numpy starts using a formatter some day, the formatter won't split the line.

charris commented on numpy/polynomial/_polybase.py:
```diff
@@ -199,12 +199,7 @@ def has_samecoef(self, other):
-            return False
-        else:
-            return True
+        return len(self.coef) == len(other.coef) and np.all(self.coef == other.coef)
```
Again, the original is easier to read.

DimitriPapadopoulos replied:
I've just split it into two lines for readability. However, if we use a formatter in the future, it would unfold again into a single line (because it wouldn't exceed the max line length).

charris commented on numpy/tests/test_public_api.py:
```diff
@@ -271,19 +271,12 @@ def test_NPY_NO_EXPORT():
+        and name not in PUBLIC_MODULES
+        and name not in PUBLIC_ALIASED_MODULES
+        and name not in PRIVATE_BUT_PRESENT_MODULES
+    )
```
I like the breaking across lines, but putting a lot of computation in the return is, I think, a bad habit to be avoided.

DimitriPapadopoulos replied:
The computation could be put in a variable, and then the variable returned. But then a `RET` rule would suggest to avoid the variable 😀

DimitriPapadopoulos replied:
I've kept this change because it's hardly a computation. I could use a temporary variable, but that's probably overkill.

charris commented:
This if very much a mixed bag. I think making all these changes automatic would be mistake and likely to obfuscate some code. Hand picking some of the better ones would be good, but on balance, taking all of them would be a bit much. I would class these as suggestions, not requirements.

DimitriPapadopoulos commented:
Let's try to disable these rules (perhaps keeping the few changes that really seem to improve the existing code base):
* SIM102
* SIM103 (not because it is broken but because automatically applying it doesn't seem to improve the code much)
* SIM114

SIM401 clearly improves the situation, so I'll keep it.

What about SIM910, changing `.get(..., None)` to `.get(...)`? I do prefer it for the reasons exposed in my previous answer, but that's not everyone's taste.

DimitriPapadopoulos committed:
MAINT: Apply ruff/flake8-simplify rule SIM103

SIM103 Return the condition directly
SIM103 Return the negated condition directly

DimitriPapadopoulos committed:
MAINT: Apply ruff/flake8-simplify rule SIM905

SIM905 Consider using a list literal instead of `str.split`

DimitriPapadopoulos commented:
* I have rebased.
* I have removed SIM102 and SIM114 changes, and also removed SIM114 chanegs except a couple that seem to make sense to me.
* I have kept SIM101, SIM401 and SIM910 changes.
* I have added SIM905 changes.

DimitriPapadopoulos commented on doc/source/conf.py:
```diff
@@ -432,14 +432,11 @@ def setup(app):
-    """.split()
-coverage_ignore_classes = r"""
-    """.split()
+coverage_ignore_modules = []
```
This is strange. Was the intent really to create an empty array? Why use `r"""    """.split()` instead of an empty array?

charris replied:
That code is from 2008 :) I suspect it was left that way as a sort of template for future additions. Why not use a list for the others? The only thing I can think of is that it was easy to do, but who knows. The change looks fine.

charris commented:
Thanks @DimitriPapadopoulos .
